plugins/woocommerce/includes/class-wc-order.php

```diff
@@ -103,25 +103,25 @@ public function payment_complete( $transaction_id = '' ) {
 		}
 
 		try {
-			do_action( 'woocommerce_pre_payment_complete', $this->get_id() );
+			do_action( 'woocommerce_pre_payment_complete', $this->get_id(), $transaction_id );
 
 			if ( WC()->session ) {
 				WC()->session->set( 'order_awaiting_payment', false );
 			}
 
-			if ( $this->has_status( apply_filters( 'woocommerce_valid_order_statuses_for_payment_complete', array( 'on-hold', 'pending', 'failed', 'cancelled' ), $this ) ) ) {
+			if ( $this->has_status( apply_filters( 'woocommerce_valid_order_statuses_for_payment_complete', array( 'on-hold', 'pending', 'failed', 'cancelled' ), $this, $transaction_id ) ) ) {
 				if ( ! empty( $transaction_id ) ) {
 					$this->set_transaction_id( $transaction_id );
 				}
 				if ( ! $this->get_date_paid( 'edit' ) ) {
 					$this->set_date_paid( time() );
 				}
-				$this->set_status( apply_filters( 'woocommerce_payment_complete_order_status', $this->needs_processing() ? 'processing' : 'completed', $this->get_id(), $this ) );
+				$this->set_status( apply_filters( 'woocommerce_payment_complete_order_status', $this->needs_processing() ? 'processing' : 'completed', $this->get_id(), $this, $transaction_id ) );
```

Comment:
Despite our initial suggestion and, similar to the above, adding `$transaction_id` can be a bit problematic as there are other places where this filter is also applied but with no `$transaction_id` available. (See [here](https://github.com/woocommerce/woocommerce/blob/542f7ef91850039ac92018432e7a24bb6aa1adf3/plugins/woocommerce/includes/class-wc-order.php#L292) and [here](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/data-stores/class-wc-order-data-store-cpt.php#L159)).

This might very well end up being a non-issue or we could even think about this parameter as optional, but it feels "wrong" that the result of a filter might depend on a parameter that is sometimes available and sometimes isn't. This could lead to different behavior depending on where and when it is called, which is not ideal.

If not really necessary, I think we should refrain from adding the parameter to the filters, but keep them for the actions. Would that work for your use case?